### PR TITLE
some tweaks

### PR DESCRIPTION
**Pull request by sfuerte**
- added cloud.id & auth fields
- support for prefix removal when root field flag is on
- support for pre-built modules (that come with DEB package)
- Hiera usage example

### CODE REVIEW

**Comment by corey-hammerton** [comment]
Please resolve merge conflicts and build errors before this can be merged. Advise you use Puppet Development Kit.

**Comment by sfuerte** [comment]
will check a bit later. Though I'm wondering why 2494ee1 began to crash to the previous ones were fine. That commit is just few lines in README.

File: manifests/config.pp [hunk 1]
```diff
@@ -5,11 +5,26 @@
...
+
+  # check if array is empty, no need to create a config entry then
+  if $metricbeat::modules[0].length() > 0 {
+    $modules_arr = $metricbeat::modules
+  } else {
+    $modules_arr = undef
+  }
+
+  # if fields are "under root", then remove prefix
+  if $fields_under_root == true {
```
Comment (corey-hammerton) [inline manifests/config.pp]: Please refer to this variable as $metricbeat::fields_under_root. This is one of the causes for the build failures.

**Comment by sfuerte** [comment]
the merge conflict comes from 4c030121521e558762d189769398afef50998abd that got committed just 10 mins ago. Will re-do in that format a bit later. 
Also, I did rename `config_file` to `config_dir` with the consecutive changes: `metricbeat.yml` filename is pretty much a constant and `modules.d` depends on its folder, so it's more logical to define the folder itself and not the full path of the former.

**Comment by sfuerte** [comment]
> resolve build errors 

it looks like the CI file needs to be verified and updated if required. Looks like there is a conflict in versions (based on `test config` command format). 
Also, for whatever reason `windows-2012 R2-x64` got resolved as Linux and not Windows.

```
metricbeat
  on sles-12-x86_64
    should compile into a catalogue without dependency cycles
  on centos-7-x86_64
    should compile into a catalogue without dependency cycles (FAILED - 1)
    metricbeat::config
    should compile into a catalogue without dependency cycles
  on windows-2012 R2-x64
      should contain File[metricbeat.yml] with ensure => "present", owner => "root", group => "root", mode => "0600", path => "/etc/metricbeat/metricbeat.yml" and validate_cmd => "/usr/share/metricbeat/bin/metricbeat -configtest -c %" (FAILED - 2)
      with ensure = absent
    should compile into a catalogue without dependency cycles
  on ubuntu-18.04-x86_64
        should contain File[metricbeat.yml] with ensure => "absent", path => "/etc/metricbeat/metricbeat.yml" and validate_cmd => "/usr/share/metricbeat/bin/metricbeat -configtest -c %" (FAILED - 3)
      with disable_configtest = true
    should compile into a catalogue without dependency cycles
  on oraclelinux-7-x86_64
        should contain File[metricbeat.yml] with ensure => "present", owner => "root", group => "root", mode => "0600", path => "/etc/metricbeat/metricbeat.yml" and validate_cmd defined (FAILED - 4)
      with config_mode = 0644
    should compile into a catalogue without dependency cycles
  on windows-2012-x64
    should compile into a catalogue without dependency cycles
  on debian-9-x86_64
        should contain File[metricbeat.yml] with ensure => "present", owner => "root", group => "root", mode => "0644", path => "/etc/metricbeat/metricbeat.yml" and validate_cmd => "/usr/share/metricbeat/bin/metricbeat -configtest -c %" (FAILED - 5)
      with config_mode = 9999
    should compile into a catalogue without dependency cycles
  on fedora-25-x86_64
        should raise Puppet::Error
      with major_version = 6 for new config test flag
    should compile into a catalogue without dependency cycles
  on scientific-7-x86_64
        should contain File[metricbeat.yml] with ensure => "present", owner => "root", group => "root", mode => "0600", path => "/etc/metricbeat/metricbeat.yml" and validate_cmd => "/usr/share/metricbeat/bin/metricbeat --path.config /etc/metricbeat test config" (FAILED - 6)
    metricbeat::install
    should compile into a catalogue without dependency cycles
```

**Comment by corey-hammerton** [comment]
See https://travis-ci.org/corey-hammerton/puppet-metricbeat/jobs/493783919 for more details. But some of the build errors are show as follows:

```  1) metricbeat on sles-12-x86_64 metricbeat::config with major_version = 6 for new config test flag should contain File[metricbeat.yml] with ensure => "present", owner => "root", group => "root", mode => "0600", path => "/etc/metricbeat/metricbeat.yml" and validate_cmd => "/usr/share/metricbeat/bin/metricbeat --path.config /etc/metricbeat test config"
     Failure/Error:
       is_expected.to contain_file('metricbeat.yml').with(
         ensure: 'present',
         owner: 'root',
         group: 'root',
         mode: '0600',
         path: '/etc/metricbeat/metricbeat.yml',
         validate_cmd: '/usr/share/metricbeat/bin/metricbeat --path.config /etc/metricbeat test config',
       )
     
       expected that the catalogue would contain File[metricbeat.yml] with validate_cmd set to "/usr/share/metricbeat/bin/metricbeat --path.config /etc/metricbeat test config" but it is set to "/usr/share/metricbeat/bin/metricbeat --path.config ${metricbeat::config_dir} test config"
     # ./spec/classes/metricbeat_spec.rb:115:in `block (6 levels) in <top (required)>'
  2) metricbeat on sles-12-x86_64 with ensure = absent should compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.to compile }
       dependency cycles found: (File[/etc/metricbeat/modules.d/system.yml] => Service[metricbeat] => Class[Metricbeat::Service] => Class[Metricbeat::Install] => Package[metricbeat] => File[/etc/metricbeat/modules.d/system.yml])
     # ./spec/classes/metricbeat_spec.rb:386:in `block (5 levels) in <top (required)>'
  3) metricbeat on sles-12-x86_64 with multiple modules should compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.to compile }
       dependency cycles found: (File[/etc/metricbeat/modules.d/system.yml] => Service[metricbeat] => Class[Metricbeat::Service] => Class[Metricbeat::Install] => Package[metricbeat] => File[/etc/metricbeat/modules.d/system.yml])
     # ./spec/classes/metricbeat_spec.rb:417:in `block (5 levels) in <top (required)>'
  4) metricbeat on sles-12-x86_64 with multiple processors should compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.to compile }
       dependency cycles found: (File[/etc/metricbeat/modules.d/system.yml] => Service[metricbeat] => Class[Metricbeat::Service] => Class[Metricbeat::Install] => Package[metricbeat] => File[/etc/metricbeat/modules.d/system.yml])
     # ./spec/classes/metricbeat_spec.rb:433:in `block (5 levels) in <top (required)>'```

**Comment by corey-hammerton** [comment]
As I mentioned in one of the reviews, please change references from `Package['metricbeat']` to `Class['metricbeat::install']` and `Service['metricbeat']` to `Class['metricbeat::service']`

**Comment by corey-hammerton** [comment]
Also, resource ordering is handled by chaining in `init.pp`. There's no reason to have `require` and `notifies` attributes in specific resources.

**Comment by sfuerte** [comment]
build errors have been fixed.

**Pull request merged**